core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -456,6 +456,31 @@ class TransactionsTest extends KafkaServerTestHarness {
   }
 
   @Test
+  def testFencingOnTransactionExpiration(): Unit = {
+    val producer = createTransactionalProducer("expiringProducer", transactionTimeoutMs = 100)
+
+    producer.initTransactions()
+    producer.beginTransaction()
+
+    // The first message and hence the first AddPartitions request should be successfully sent.
+    val firstMessageResult = producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "1", "1", willBeCommitted = false)).get()
+    assertTrue(firstMessageResult.hasOffset)
+
+    // Wait for expiration cycle to kick in.
+    Thread.sleep(600)
```

Comment:
Could we use a mock time for unit test? Having thread.sleep would introduce another flakiness.

Reply:
This is an integration test, not a unit test. So we can't use mock time, as there are no mocks. We are actually verifying that the first record was eventually aborted and the second one was never allowed to be written from end to end (ie. producer to consumer).